Fix performance problem with Split course import/export.

doctoryes opened the pull request:
* Add BlockData class to wrap all block data loaded from the Split modulestore, in order to separate serializable block data from out-of-band, non-saved, temporary metadata. This PR changes the BlockData class from a dict-derived class to an actual object with specific attributes.
* Add EditInfo class to encapsulate a block's editing & version info.
* Eliminate the deepcopy causing performance problems with import/export.
* Fix cleanup of test MongoDB modulestores when nesting modulestore creation.
* Improve bulk assertions by adding file & line number to each assertion.

https://openedx.atlassian.net/browse/PLAT-416

doctoryes commented:
@cpennington  For your review. Is my Python assumption correct: in my class derived from `dict`, the attribute in `BlockData` will *not* get serialized with the `dict`? I'm doubting it - I think I've been misled by what I see when I print the object in the debugger.

cpennington commented:
I would tend to not subclass dict, but instead have a class with `fields`, `block_type`, `definition

cpennington commented:
I'd also just use the definition_loaded attribute directly, rather than having methods to read/set it.

doctoryes commented:
@cpennington I've re-worked this PR as you suggested. The creation points for `BlockData` objects are `structure_from_mongo` and `_new_block`. What do you think?

cpennington commented on common/lib/xmodule/xmodule/modulestore/split_mongo/__init__.py:
```diff
@@ -21,3 +21,78 @@ def from_usage_key(cls, usage_key):
+        delattr(self, key)
+
+    def __iter__(self):
+        return self.__dict__.iterkeys()
```
Any reason to keep these dict-like accessors? Are there a lot of places in the code that currently use them? If not, it seems better to just take them out (no need for the object to have both attributes and keys for the same data).

doctoryes replied:
I left them in order to disturb the rest of the code as little as possible *and* because I wasn't sure that I knew all the stored attributes. But - I agree that it's unnecessary to fake a dict while using an object - goes against our "Make it simple!" mantra.

I'll change it.

doctoryes commented:
@cpennington  OK - tests now pass. This PR is ready for a re-review.
@jimabramson You were lucky enough to be tangentially related to this PR - review, please?

cpennington commented on common/lib/xmodule/xmodule/modulestore/split_mongo/__init__.py:
```diff
@@ -52,22 +112,33 @@ def from_storable(self, stored):
+    def __contains__(self, item):
+        return hasattr(self, item)
 
     def __getitem__(self, key):
```
Is this needed anymore?

doctoryes replied:
(See below...)

cpennington commented on common/lib/xmodule/xmodule/modulestore/split_mongo/__init__.py:
```diff
@@ -52,22 +112,33 @@ def from_storable(self, stored):
+                "defaults={0.defaults}, "
+                "edit_info={0.edit_info})").format(self)
+
+    def __contains__(self, item):
```
Why do we need this?

doctoryes replied:
The problem method is `ModuleStoreRead`'s `_block_matches`/`_is_set_on` - it has a `not in` that generalizes the search for fields/attributes. It's unclear to me if there's another way to do this generally - I could possibly break it out into XBlock-specific and BlockData-specific code.

cpennington replied:
The two case (field or XBlock) seem like they could be handled separately. In particular, if we have an XBlock, then we use `xblock.fields`, and if we have a BlockData, then we use `block_data.fields`. In each case, we get a dict that we can use `in` to check against.

In fact, doing it that way makes the code even smaller, because you're calling `.fields` in both branches (rather than only in one like it does now).

doctoryes replied:
I can't call `.fields` in both cases  - because of the `if` clause here:
https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/modulestore/split_mongo/split.py#L1033
It checks the top-level against `qualifiers` and the `fields`-level against `settings`. Sometimes the `BlockData` object is passed-in and sometime `BlockData.fields` is passed-in.

jimabramson commented:
@doctoryes nice change.  a question, why was `__slots__` added and then removed ?

jimabramson commented on common/lib/xmodule/xmodule/modulestore/split_mongo/__init__.py:
```diff
@@ -23,6 +23,66 @@ def from_usage_key(cls, usage_key):
+            'original_usage_version': self.original_usage_version,
+        }
+
+    def from_storable(self, edit_info):
```
trivial / personal taste thing, but this could be implemented as a classmethod which could be used in concert with a more passive/straightforward `__init__` method that took kwargs or explicit args for everything.

doctoryes replied:
The method to my madness here was to make `to_storable`/`from_storable` symmetric methods that emphasize the attributes that persist. I reserved `__init__` for the attributes that were present but did not persist.

But I agree that I should either use `kwargs` or spell out all args. I'll switch it to `kwargs`.

doctoryes commented:
@jimabramson - @cpennington and I decided that using `__slots__` was a bit too draconian. Since the code now explicitly controls what values are serialized and which aren't, it doesn't matter if `BlockData` users add arbitrary attributes to the object.

jimabramson commented:
:+1: lgtm

cpennington commented:
otherwise, :+1:

jimabramson commented:
:+1: aside from my last nitpick

jimabramson commented:
:+1: :+1:

doctoryes commented:
@cpennington  My last commit gets rid of both `__contains__` and `__getitem__` from `BlockData`. Would you like to review? Other than that, this PR is ready to squash and merge.

doctoryes commented:
@cpennington  I've changed the arg name and updated the comments for `_block_matches`. The way that it currently operates is the way the code is expected to work. I could remove the `__dict__` by special-casing the `BlockData` case - but this seems better than duplicating the comparison logic elsewhere.

doctoryes commented:
@cpennington and I discussed this final issue. The `_block_matches` method still isn't perfect - but we've deemed it good enough.

doctoryes commented:
test this please
